### PR TITLE
INIT SpeedTest

### PR DESCRIPTION
**Pull request by kunal-511**
[SpeedTest] INIT

### CODE REVIEW

**Comment by kunal-511** [comment]
Please review this PR request

File: speed test/main.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		if err != nil {
+			log.Println("Upload test failed for server:", s.Name)
+			continue
+		}
+
+		result := SpeedResult{
+			DownloadSpeed: formatSpeed(s.DLSpeed),
+			UploadSpeed:   formatSpeed(s.ULSpeed),
+		}
+		lesgo = append(lesgo, result)
```
Comment (Pushkarm029) [inline speed test/main.go:61]: Variable name can be improved.

File: speed test/main.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/showwin/speedtest-go/speedtest"
+)
+
+type SpeedResult struct {
+	DownloadSpeed string `json:"download_speed"`
+	UploadSpeed   string `json:"upload_speed"`
+}
+
+func SpeedtestHandler(w http.ResponseWriter, r *http.Request) {
+	var speedtestClient = speedtest.New()
```
Comment (Pushkarm029) [inline speed test/main.go:18]: Please consider factory type architecture.

File: speed test/main.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/showwin/speedtest-go/speedtest"
```
Comment (Pushkarm029) [inline speed test/main.go:9]: Add this comment here

TODO: Remove this third party library later

File: speed test/main.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+}
+
+func formatSpeed(speed float64) string {
+	return fmt.Sprintf("%.2f Mbps", speed)
+}
+
+func main() {
+	http.HandleFunc("/speedtest", SpeedtestHandler)
+	log.Fatal(http.ListenAndServe(":8080", nil))
```
Comment (Pushkarm029) [inline speed test/main.go:81]: Use Uber Zap instead

File: speed test/main.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	for _, s := range targets {
+		// Ping test
+		err := s.PingTest(nil)
+		if err != nil {
+			log.Println("Ping test failed for server:", s.Name)
+			continue
+		}
+
+		// Download test
+		err = s.DownloadTest()
```
Comment (Pushkarm029) [inline speed test/main.go:44]: Can we add tests?

**Review by Pushkarm029 — CHANGES_REQUESTED** [review]

**Comment by kunal-511** [comment]
@Pushkarm029  I have resolved all the issues raised by you now. But test is missing I will add them shortly

**Comment by Pushkarm029** [comment]
@kunal-511 good

**Comment by kunal-511** [comment]
Test cases have now been added.
Please merge this also @Pushkarm029

File: speedTest/main.go [hunk 6]
```diff
@@ -102,6 +102,68 @@ func formatSpeed(speed float64) string {
 	return fmt.Sprintf("%.2f Mbps", speed)
 }
 
+func TestSpeedtestHandler(t *testing.T) {
```
Comment (Pushkarm029) [inline speedTest/main.go:105]: move this into a new file `package_test.go`

File: speedTest/main.go [hunk 7]
```diff
@@ -102,6 +102,68 @@ func formatSpeed(speed float64) string {
...
+		t.Errorf("failed to decode JSON response: %v", err)
+	}
+
+	// Check if the response contains valid speed results
+	if len(results) == 0 {
+		t.Error("handler returned empty result set")
+	}
+}
+
+func TestFormatSpeed(t *testing.T) {
```
Comment (Pushkarm029) [inline speedTest/main.go:143]: same

File: speedTest/main.go [hunk 8]
```diff
@@ -102,6 +102,68 @@ func formatSpeed(speed float64) string {
...
+		result := formatSpeed(test.speed)
+		if result != test.expected {
+			t.Errorf("formatSpeed(%f) returned %s, expected %s", test.speed, result, test.expected)
+		}
+	}
+}
+
+func TestMain(t *testing.T) {
+	// simple test case to check if the main function runs without error
+	go main()
```
Comment (Pushkarm029) [inline speedTest/main.go:164]: why starting goroutine?

**Review by Pushkarm029 — CHANGES_REQUESTED** [review]

**Review by Pushkarm029 — APPROVED** [review]

**Comment by Pushkarm029** [comment]
Thanks🎉

**Pull request merged**